Riot/Utils/FormattedBodyParser.swift

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2020 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Foundation
+
+@objcMembers
+final class FormattedBodyParser: NSObject {
+    
+    private var formattedBody: String
+    
+    /// Initialize parser
+    /// - Parameter formattedBody: html formatted body
+    init(withFormattedBody formattedBody: String) {
+        self.formattedBody = formattedBody
+    }
+    
+    /// Gets visible url for a given url. Assumes formattedBody has a link like: '<a href="https://example.com/given">https://example.com/visible</a>'
+    /// - Parameter url: the url given as target
+    /// - Returns: visible url if found, otherwise nil
+    func getVisibleURL(forURL url: URL) -> URL? {
```

Comment:
You will miss this edge case if you look only at the first link occurence: '<a href="https://example.com/given">https://example.com/given</a> some text <a href="https://example.com/given">https://example.com/visible</a>'

You can copy the regex `kHTMLATagRegexPattern` from `MXKEventFormatter` to use it in `getVisibleURL`.
I think that we can even have something like this to return all visible occurrences:
```
struct HTMLURLAnchor {
    let link: String
    let content: String
}

final class FormattedBodyParser: NSObject {

private enum Constants {
    static let htmlURLAnchorTagRegexPattern = "<a href=\"(.*?)\">([^<]*)</a>"
}

let htmlURLAnchorRegex: NSRegularExpression

func getHTMLURLAnchors() -> [HTMLURLAnchor] {
    // Use regex here self.htmlURLAnchorRegex
    // build and return list with `HTMLURLAnchor`
}

func getVisibleURLs(forURL url: URL) -> URL? {
    self.getHTMLAnchorsWithURL().map { htmlURLAnchor in
        return htmlURLAnchor.content
    }
}

}
```

Reply:
Changed using the regex, but it doesn't seem possible to distinguish the edge case scenario without passing the url range to the delegate method.

> guard let contentURL = URL(string: htmlURLAnchor.content), url == contentURL else {
>             return nil
>         }

Also we can't use above `url == contentURL`. If we do so, we lose the items having different link and contents. But we need to extract the visible url for them.